### PR TITLE
Numeric type audit: network, consensus, provider, rpc-types

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Ref #433

## Solution

* `gas_limit` -> `u128`
* `gas` -> `u128`
* `gas_price` -> `u128`
* `max_fee_per_blob_gas` -> `u128`
* `transaction_type` -> `u8`
* `max_fee_per_gas` -> `u128`
* `max_priority_fee_per_gas` -> `u128`
* Consequently, changed fn args and return values accordingly.
* Satisfies clippy, tests will fail.

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [x] Breaking changes

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
I think all of these make sense, primitive types are preferable imo.
gas being u128 should make some operations easier but should not affect math for any call arguments etc

File: crates/provider/src/layers/gas.rs [hunk 1]
```diff
@@ -81,13 +80,15 @@ where
...
         self.inner.get_gas_price().await
     }
 
     /// Gets the gas_limit to be used in txs.
-    async fn get_gas_estimate(&self, tx: &N::TransactionRequest) -> TransportResult<U256> {
-        self.inner.estimate_gas(tx, None).await
+    async fn get_gas_estimate(&self, tx: &N::TransactionRequest) -> TransportResult<u128> {
+        self.inner.estimate_gas(tx, None).await.map(|gas| gas.to::<u128>()) // Temp Fix until
+                                                                            // estimate_gas returns
+                                                                            // u128
```
Comment (Evalir) [inline crates/provider/src/layers/gas.rs:91]: Should this be solved on the same PR?

File: crates/provider/src/provider.rs [hunk 2]
```diff
@@ -806,15 +813,17 @@ pub trait Provider<T: Transport + Clone = BoxTransport, N: Network = Ethereum>:
...
         &self,
-        block_count: U256,
+        block_count: u64,
         last_block: BlockNumberOrTag,
         reward_percentiles: &[f64],
     ) -> TransportResult<FeeHistory> {
         self.client().request("eth_feeHistory", (block_count, last_block, reward_percentiles)).await
     }
 
+    // TODO: Return TransportResult<u128> instead of U256
```
Comment (Evalir) [inline crates/provider/src/provider.rs:826]: same here

File: crates/serde/src/num.rs [hunk 3]
```diff
@@ -264,6 +264,33 @@ pub mod u128_hex_or_decimal {
     }
 }
 
+/// serde functions for handling primitive optional `u128` as [U128](alloy_primitives::U128)
+pub mod u128_hex_or_decimal_opt {
+    // TODO: Add unit tests for this module
```
Comment (Evalir) [inline crates/serde/src/num.rs:269]: Also here

**Review by Evalir — COMMENTED** [review]
This mostly looks good to me, I'd just like to ensure we get the TODOs taken care of and add enough tests to be 100% sure we're not breaking serde

File: crates/rpc-types/src/eth/transaction/tx_type.rs [hunk 4]
```diff
@@ -45,6 +45,7 @@ impl From<TxType> for u8 {
     }
 }
 
+// TODO(Numeric-Audit): Remove this conversion
```
Comment (gakonst) [inline crates/rpc-types/src/eth/transaction/tx_type.rs:48]: todo

File: crates/provider/src/utils.rs [hunk 5]
```diff
@@ -46,10 +46,14 @@ pub fn eip1559_default_estimator(
     base_fee_per_gas: U256,
     rewards: &[Vec<U256>],
 ) -> Eip1559Estimation {
+    // TODO: Num type audit - return u128 here
```
Comment (gakonst) [inline crates/provider/src/utils.rs:49]: todo

**Review by gakonst — CHANGES_REQUESTED** [review]

File: crates/consensus/src/transaction/eip2930.rs [hunk 6]
```diff
@@ -29,8 +29,8 @@ pub struct TxEip2930 {
     /// this transaction. This is paid up-front, before any
     /// computation is done and may not be increased
     /// later; formally Tg.
-    #[cfg_attr(feature = "serde", serde(with = "alloy_serde::u64_hex_or_decimal"))]
-    pub gas_limit: u64,
+    #[cfg_attr(feature = "serde", serde(with = "alloy_serde::u128_hex_or_decimal"))]
```
Comment (prestwich) [inline crates/consensus/src/transaction/eip2930.rs:32]: do any endpoints use decimal here? in other PRs we've avoided allowing decimal when nobody actually uses it

File: crates/rpc-types/src/eth/transaction/mod.rs [hunk 7]
```diff
@@ -42,29 +42,48 @@ pub struct Transaction {
...
     pub to: Option<Address>,
     /// Transferred value
     pub value: U256,
     /// Gas Price
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub gas_price: Option<U256>,
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        with = "alloy_serde::num::u128_hex_or_decimal_opt"
```
Comment (prestwich) [inline crates/rpc-types/src/eth/transaction/mod.rs:60]: same question. pretty sure decimal would be illegal rpc response here? I don't mind tolerating it, but @mattsse asked for removal of decimal in other PRs

**Review by prestwich — APPROVED** [review]
just style nit for @mattsse

File: crates/rpc-types/src/eth/transaction/mod.rs (hunk 7 above)
Comment (Evalir) [reply crates/rpc-types/src/eth/transaction/mod.rs:60]: + 1 to removing decimal support

Comment (mattsse) [reply crates/rpc-types/src/eth/transaction/mod.rs:60]: makes sense, let's enforce strict quantity rules everywhere for core eth types

Comment (prestwich) [reply crates/rpc-types/src/eth/transaction/mod.rs:60]: side note, the `U***` type deser DO accept decimal. so this is technically a breaking behavior change

**Comment by prestwich** [comment]
followup work: #465

**Pull request merged**